Plutarch/Builtin.hs

```diff
@@ -258,10 +260,28 @@ instance PIsData PBool where
       nil :: Term s (PBuiltinList PData)
       nil = pnil
 
+-- | NB: `PAsData (PBuiltinPair (PAsData a) (PAsData b))` and `PAsData (PTuple a b)` have the same representation.
+instance PIsData (PBuiltinPair (PAsData a) (PAsData b)) where
+  pfromData x = f # x
+    where
+      f = phoistAcyclic $
+        plam $ \pairDat -> plet (psndBuiltin #$ pasConstr # pforgetData pairDat) $
+          \pd -> ppairDataBuiltin # punsafeCoerce (phead # pd) #$ punsafeCoerce (phead #$ ptail # pd)
+  pdata x = punsafeCoerce target
+    where
+      target :: Term _ (PAsData (PBuiltinPair PInteger (PBuiltinList PData)))
+      target = f # punsafeCoerce x
+      f = phoistAcyclic $
+        plam $ \pair -> pconstrBuiltin # 0 #$ pcons # (pfstBuiltin # pair) #$ pcons # (psndBuiltin # pair) # pnil
+
+instance PIsData PUnit where
+  pfromData _ = pconstant ()
+  pdata _ = punsafeCoerce $ pconstant (Constr 0 [])
```

Comment:
Why not an empty list as the internal representation?

Reply:
Have to follow the rules set before us. `toData ()` from Plutus creates a `Constr 0 []`.

Reply:
`pfromData . pconstantData`, but this is valid in either case in this case! In any case I agree yeah.